From 3b9fed51e91c230dc2440266039be12138e33d76 Mon Sep 17 00:00:00 2001
From: Juergen Hoetzel <juergen@archlinux.org>
Date: Wed, 6 Oct 2010 12:53:52 +0200
Subject: [PATCH 1/2] use java.nio in do-copy method for Files

NIO Channels reduce CPU/Disk load when copying Files (by using
syscalls like sendfile internally on Linux/Solaris).
---
 src/clj/clojure/java/io.clj |    7 ++++---
 1 files changed, 4 insertions(+), 3 deletions(-)

diff --git a/src/clj/clojure/java/io.clj b/src/clj/clojure/java/io.clj
index 4279569..a650f4f 100644
--- a/src/clj/clojure/java/io.clj
+++ b/src/clj/clojure/java/io.clj
@@ -334,9 +334,10 @@
     (do-copy in output opts)))
 
 (defmethod do-copy [File File] [#^File input #^File output opts]
-  (with-open [in (FileInputStream. input)
-              out (FileOutputStream. output)]
-    (do-copy in out opts)))
+  (with-open [in (-> input FileInputStream. .getChannel)
+              out (-> output FileOutputStream. .getChannel)]
+    (.transferTo in 0 (.size in) out))
+  nil)
 
 (defmethod do-copy [String OutputStream] [#^String input #^OutputStream output opts]
   (do-copy (StringReader. input) output opts))
-- 
1.7.3.2

